drafts/0007-RFC-BAG-OPERATORS.md

```diff
@@ -0,0 +1,564 @@
+
+### Example Environment
+
+TODO: I would like some feedback on this example data. In the real world, we don't always control the shape of our data; but a reader may look at this and think "why isn't this flat if employees is homogenous?". As a database schema, this would be an "Employees" table and a "Department" table with a separate relation table. I would like some feedback on example data shape that would avoid this initial reaction.
+
+```
+{ 
+  'hr': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Bob Smith',   'title': 'Consultant' }, 
+      { 'id': 2, 'name': 'Susan Smith', 'title': 'Manager' },
+      { 'id': 3, 'name': 'Jane Smith',  'title': 'DEI'}
+    >>
+  },
+  'accounting': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Quincy Jones',   'title': 'Director' }, 
+      { 'id': 2, 'name': 'James Early Jones', 'title': 'Manager' },
+      { 'id': 3, 'name': 'Indiana Jones',  'title': 'Intern'}
+    >>
+  },
+  'engineering': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Paul McCharmly',   'title': 'Manager' }, 
+      { 'id': 2, 'name': 'George Parasol', 'title': 'SDE' },
+      { 'id': 3, 'name': 'Ringo Stone',  'title': 'Principal SDE'},
+      { 'id': 4, 'name': 'Eric Lennon',  'title': 'Intern'}
+    >>
+  }
+} 
+```
+
+### Union Examples
+
+**Union 1** — No manager meeting invite list
+```
+PartiQL> SELECT id, name FROM (
+   |   SELECT * FROM hr.employees
+   |   UNION
+   |   SELECT * FROM accounting.employees
+   |   UNION
+   |   SELECT * FROM engineering.employees
+   | )
+   | WHERE title != 'Manager'
+   | ORDER BY title ASC
+   |
+==='
+[
+  {
+    'id': 1,
+    'name': 'Bob Smith'
+  },
+  {
+    'id': 3,
+    'name': 'Jane Smith'
+  },
+  {
+    'id': 1,
+    'name': 'Quincy Jones'
+  },
+  {
+    'id': 3,
+    'name': 'Indiana Jones'
+  },
+  {
+    'id': 4,
+    'name': 'Eric Lennon'
+  },
+  {
+    'id': 3,
+    'name': 'Ringo Stone'
+  },
+  {
+    'id': 2,
+    'name': 'George Parasol'
+  }
+]
+---
+OK!
+```
+
+**Union 2** — Union where right operands are expressions rather than SFW blocks. Note that this is stated as possible in the SQL++ text, but the grammar does not support this. This is the result of an initial implementation because our parser actually parsers UNION operators as expressions rather than SFW blocks as described by the grammar.
+
+```
+PartiQL> SELECT name FROM (
+   |   SELECT * FROM hr.employees UNION engineering.employees UNION accounting.employees
+   | ) WHERE title = 'Intern'
+   |
+==='
+<<
+  {
+    'name': 'Eric Lennon'
+  },
+  {
+    'name': 'Indiana Jones'
+  }
+>>
+---
+OK!
+```
+
+**Union 3** -- Get all distinct titles
+```
+PartiQL> SELECT DISTINCT title FROM (
+   |   SELECT * FROM hr.employees
+   |   UNION
+   |   SELECT * FROM accounting.employees
+   |   UNION
+   |   SELECT * FROM engineering.employees
+   | )
+   | ORDER BY title ASC
```

Comment:
My recollection is there is some weird associativity with the set operators and things like `ORDER BY` and interaction with schema as well (same arity of columns, same data types, different names).  I think we need to have a SQL compatibility section about how that will work.